src/lib.rs

```diff
@@ -0,0 +1,127 @@
+#[deny(clippy::all)]
+///! Graph API Traits
+///
+use rand::SeedableRng;
+
+#[allow(dead_code)]
+/// A graph layer name.
+struct Layer(&'static str);
+
+/// Abstract object in a graph, eg. node or edge.
+trait GraphObject {
+    /// Identifier of all graph objects.
+    type Id;
+
+    /// Local graph object data.
+    type Data;
+
+    /// Return the edge id.
+    fn id(&self) -> Self::Id;
+
+    /// Get local data.
+    fn data(&self) -> &Self::Data;
+
+    /// Get mutable local data.
+    fn data_mut(&self) -> &mut Self::Data;
+}
+
+/// A graph  node.
+trait Node: GraphObject {
+    /// The edge type connected to this node.
+    type Edge: GraphObject;
+
+    /// Get connected nodes.
+    fn neighbors(&self) -> Iterator<Item = &Self>;
+
+    /// Get incoming and outgoing edges.
+    fn edges(&self) -> Iterator<Item = &Self::Edge>;
+}
+
+/// A graph edge between two nodes.
+trait Edge: GraphObject {
+    /// Get the edge weight.
+    fn weight(&self) -> f64;
+
+    /// The object from which the edge starts.
+    fn from(&self) -> <Self as GraphObject>::Id;
+
+    /// The object to which the edge points.
+    fn to(&self) -> <Self as GraphObject>::Id;
```

Comment:
This feels iffy -- you are saying that an edge has as a source and destination an `Edge::Id`, but this is not true, that's actually a `Node::Id`. If it happens that the `Id` for both nodes & edges coincide this is obviously correct, but I would say very confusing from an interface perspective.

The fact in our _concrete_ case both ids will be some sort of hash it's an implementation detail specific to our business domain, but if feels a bit strange to be exploiting it. You could of course lock in place the `Id` and make it monomorphic and not an associated type of the `GraphObject`, but doing so means you cannot test things easily, where you want the notion of some mock ID.

Reply:
The idea is that Nodes and Edges share the same type for Id. In my understanding, this is how you currently have it in the GraphAPI.

Maybe my implementation doesn't guarantee that though :thinking:

Reply:
> The idea is that Nodes and Edges share the same type for Id. In my understanding, this is how you currently have it in the GraphAPI.

Yes, but in my implementation you cannot construct an edge where <source,target> maps to the wrong Id type. In other terms, this is what you can write in your `Edge` trait:

```rust
pub struct MyNode {
  id: u32
}

pub struct MyEdge {
  id: String
}

impl GraphObject for MyNode {
    type Id = u32
    ...
}

impl GraphObject for MyEdge {
    type Id = String
}

impl Edge for MyEdge {
    fn from(&self) -> <Self as GraphObject>::Id; // ~ String
}
```

Do you see the problem? Now `Id` has type `String`, which is incorrect as that should have been `u32`. If you compare my implementation, instead, afaik it doesn't suffer from the same problem 🤔

Reply:
Yeah that's what I meant by lacking guarantees. Will think about it more.

Reply:
Made some changes which don't have this problem anymore I believe: 
3dd89dd